utils/rate_limiter.go

```diff
@@ -62,8 +65,9 @@ func (b *rateLimiter) CheckCredit(itemCost float64) bool {
 	b.lastTick = currentTime
 	// calculate how much credit have we accumulated since the last tick
 	b.balance += elapsedTime.Seconds() * b.creditsPerSecond
-	if b.balance > b.creditsPerSecond {
-		b.balance = b.creditsPerSecond
+	upperbound := math.Max(b.creditsPerSecond, itemCost)
```

Comment:
no, to constructor. Because I think maxBalance is the fundamental property of this algorithm, and previously we just decided to set it to creditsPerSecond, which made sense but is not always right.